src/drivers/ros3.jl

```diff
@@ -4,28 +4,50 @@
 This is the read-only virtual driver that enables access to HDF5 objects stored in AWS S3
 """
 struct ROS3 <: Driver
-    fa::API.H5FD_ros3_fapl_t
-end
-
-function ROS3(
-    version::Integer,
-    auth::Bool,
-    region::AbstractString,
-    id::AbstractString,
-    key::AbstractString
-)
-    return (ROS3 ∘ API.H5FD_ros3_fapl_t)(
-        version,
-        auth,
-        Base.unsafe_convert(Cstring, region),
-        Base.unsafe_convert(Cstring, id),
-        Base.unsafe_convert(Cstring, key)
-    )
+    version::Int32
+    authenticate::Bool
+    aws_region::String
+    secret_id::String
+    secret_key::String
+    function ROS3(version, authenticate, aws_region, secret_id, secret_key)
+        length(aws_region) <= API.H5FD_ROS3_MAX_REGION_LEN ||
+            "length(aws_region), $(length(aws_region)), must be less than or equal to $(API.H5FD_ROS3_MAX_REGION_LEN)"
+        length(secret_id) <= API.H5FD_ROS3_MAX_SECRET_ID_LEN ||
+            "length(secret_id), $(length(secret_id)), must be less than or equal to $(API.H5FD_ROS3_MAX_SECRET_ID_LEN)"
+        length(secret_key) <= API.H5FD_ROS3_MAX_SECRET_KEY_LEN ||
+            "length(secret_key), $(length(secret_key)), must be less than or equal to $(API.H5FD_ROS3_MAX_SECRET_KEY_LEN)"
+        new(version, authenticate, aws_region, secret_id, secret_key)
+    end
 end
 
+ROS3() = ROS3(1, false, "", "", "")
 ROS3(region::AbstractString, id::AbstractString, key::AbstractString) =
     ROS3(1, true, region, id, key)
-ROS3() = ROS3(1, false, "", "", "")
+_ntuple_to_string(x) = unsafe_string(Ptr{Cchar}(pointer_from_objref(Ref(x))), length(x))
```

Comment:
This doesn't convert the ntuple to a string back for me
```julia
_ntuple_to_string(x) = unsafe_string(Ptr{Cchar}(pointer_from_objref(Ref(x))), length(x))
x = "Dfsddfs"
a = ntuple(i -> i <= length(x) ? Cchar(x[i]) : Cchar(0x0), 10 + 1)
julia> _ntuple_to_string(a)
"\xc8\x18\xb2\n\0\0\0\0X\xd0\xc3"
```
Shouldn't this actually look like something like the following
```julia
function _ntuple_to_string(buf)
    len = findfirst(iszero, buf) -1
    unsafe_string(Base.unsafe_convert(Ptr{Int8}, collect(buf[1:len])))
end
```

Reply:
unsafe_string already searches for a null itself.

```
julia> function _ntuple_to_string(x)
           r = Ref(x)
           GC.@preserve r unsafe_string(Ptr{Cchar}(pointer_from_objref(r)))
       end
_ntuple_to_string (generic function with 1 method)

julia> _ntuple_to_string(a)
"Dfsddfs"
```

Reply:
We also probably should use `ncodeunits` and `codeunits` like you suggested before:
https://github.com/JuliaIO/HDF5.jl/issues/818#issuecomment-780071965

Reply:
Why not use `FixedString`, which is already in the package?
https://github.com/JuliaIO/HDF5.jl/blob/0120cb2d32fc8d57727b3dcb17860c63c00a0cd3/src/typeconversions.jl#L99-L105

Reply:
FixedString is not an AbstractString.

I started working on a version of InlineStrings.jl called NStrings.jl over the weekend that basically is a full blown general implementation of this and is an AbstractString.

I'll publish that repo soon.

Reply:
I was actually about to suggest moving it to a separate repo! Nice!

Reply:
The initial version of NTuple based Strings is here:
https://github.com/mkitti/NStrings.jl

Reply:
The current NTuple based Strings are now here and registered:

https://github.com/mkitti/StaticStrings.jl